KAFKA-2017: Persist Group Metadata and Assignment before Responding

guozhangwang opened the pull request.

guozhangwang committed:
move group metadata into offset manager

guozhangwang committed:
move group metadata into offset manager

guozhangwang committed:
add back package.html

guozhangwang committed:
add back package.html

guozhangwang committed:
persist synced assignment and client side error handling

guozhangwang committed:
persist synced assignment and client side error handling

guozhangwang committed:
Add renamed files

guozhangwang committed:
Add renamed files

guozhangwang commented:
Ping @hachikuji @junrao @onurkaraman for reviews.

Major changes:
1. The offset topic message has three versions. 0 and 1 refer to offset message, and 2 refers to group metadata message.
2. Group metadata message contains the selected protocol only, with the subscription and assignment stored as byte arrays. Hence after loading each member will only have the single selected protocol stored, and future join-group requests will multiple supported protocols will be considered as new join requests.
3. When loading the topic from log upon `becomeLeader`, the background thread goes through the message, and based on the key read from the log decided to choose different manner for reading the value (offset or group metadata), and pushing them to respective caches.
4. Metadata cache depend on the underlying `Pool` for concurrent access, while offset cache has two separate locks for "range access''. One is for expiring old offsets from the scheduler thread, the other one is for clearing the cache entries upon `becomeFollower`. Offset fetches will only be guarded by the clearing cache lock. This is basically the same as the old `OffsetManager`.
5. Upon finishing the sync phase, coordinator writes the single-message into the replica manager with the ack mode, and triggers `propogateAssignment` upon completing write.
6. Coordinator check "partition loading in progress" while handling Join / Leave / OffsetCommit / OffsetFetch (we do not check upon Sync and HB), and client handle this error code as retry / ignore / retry / retry.

guozhangwang committed:
bindly accepts HB while loading in progress

guozhangwang committed:
bindly accepts HB while loading in progress

guozhangwang committed:
minor fixes

guozhangwang committed:
resolve conflicts

guozhangwang committed:
some more minor fixes on unit tests

guozhangwang committed:
fix Pool.put() call to return the group itself

guozhangwang committed:
convert between byte array and byte buffer

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+          else if (status.error == ErrorMapping.MessageSizeTooLargeCode
+            || status.error == ErrorMapping.MessageSetSizeTooLargeCode
+            || status.error == ErrorMapping.InvalidFetchSizeCode)
+            Errors.INVALID_COMMIT_OFFSET_SIZE.code
```
Is this the right error code for the SyncGroupRequest? Also, can we actually get ErrorMapping.InvalidFetchSizeCode?

guozhangwang replied:
1) InvalidFetchSizeCode actually maps to InvalidMessageSizeException in `ErrorMapping`, this can actually be thrown in produce response.

2) Ack on "Errors.INVALID_COMMIT_OFFSET_SIZE.code", will fix it.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+  /**
+   * Get the group associated with the given groupId, or null if not found
+   */
+  def getGroup(groupId: String) = {
```
It would be clearer if we explicitly define the return type.

guozhangwang replied:
Ack.

guozhangwang committed:
wip fixing CoordinatorResponseTest

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -323,8 +326,11 @@ class GroupCoordinator(val brokerId: Int,
       responseCallback(Errors.NOT_COORDINATOR_FOR_GROUP.code)
+    } else if (isCoordinatorLoadingInProgress(groupId)) {
+      // the group is still loading, so respond just blindly
+      responseCallback(Errors.NONE.code)
```
Why aren't we returning GROUP_LOAD_IN_PROGRESS for the HeartbeatResponse?

hachikuji replied:
Would the consumer handle this case any differently than NONE?

onurkaraman replied:
Maybe, maybe not. But why not let the broker be specific and let the consumer implementation decide?

hachikuji replied:
Yeah, I agree, but I can't actually think of any case where the client wouldn't just ignore it. Guess clients might want to add logging in this case?

guozhangwang replied:
I originally did the special error returning, but then I changed it to just abstracting it at the server side for the same reason @hachikuji mentioned. I feel if it is mostly the case that users will not special handle this error, we'd better simply the possible return error cases in the protocol.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+  def removeGroup(groupId: String) {
+    if (!groupsCache.contains(groupId))
+      throw new IllegalArgumentException("Cannot remove non-existing group")
+    groupsCache.remove(groupId)
```
I'm not sure if groupsCache is properly guarded. There's a check-then-act here that I'm not sure is okay.

CoordinatorMetadata used to guard its state with a global ReentrantReadWriteLock called metadataLock.

It might be better as:

``` scala
if (groupsCache.remove(groupId) == null)
  throw new IllegalArgumentException("Cannot remove non-existing group")
```

junrao replied:
Should we add a tombstone to remove the group in the offset topic?

guozhangwang replied:
Rep @onurkaraman : `removeGroup` should always be guarded by the group lock inside the `GroupCoordinator`, while `getGroup` and `addGroup` are not since the group object is not available yet. I will make that more clear in the comments.

Rep @junrao : ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+        memberStruct.set(MEMBER_METADATA_MEMBER_ID_V0, memberMetadata.memberId)
+        memberStruct.set(MEMBER_METADATA_SESSION_TIMEOUT_V0, memberMetadata.sessionTimeoutMs)
+
+        val selectedProtocol = memberMetadata.supportedProtocols.filter(_._1.equals(groupMetadata.protocol))
```
Minor: I added metadata method which does this:

``` scala
val metadata = memberMetadata.metadata(groupMetadata.protocol)
memberStruct.set(MEMBER_METADATA_SUBSCRIPTION_V0, ByteBuffer.wrap(metadata))
```

guozhangwang replied:
Cool, thanks.

junrao commented:
A couple of quick comments.
1. MemberMetadata
   Variables such as assignment, awaitingSyncCallback need to be volatile. Basically, if the reader and the writers are in different threads, we need a memory barrier for the reader to see the latest value. Setting a variable as volatile is one way to achieve that.
2. Currently, it seems that we put the group coordinator in the loading mode after the leader changes. It's possible for a broker to get a leaderAndIsr request while it's still the leader of a partition. For example, during controller failover, the new controller will resend leaderAndIsr requests for existing partitions. If the leader doesn't actually change, we don't have to reload the offset data.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+
+  def storeGroup(group: GroupMetadata,
+                 groupAssignment: Map[String, Array[Byte]],
+                 responseCallback: (GroupMetadata, Map[String, Array[Byte]], Short) => Unit) {
```
Minor: Do we need to return the group assignment? It doesn't seem like the argument ever gets modified.

onurkaraman commented:
@junrao: the current synchronization pattern maintains the invariant that a GroupMetadata is owned by one thread at a time. Any access or modifications to group state (including its members) are guarded by the GroupMetadata's implicit lock. Is this invariant being broken anywhere?

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+  }
+
+  private def addGroup(groupId: String, group: GroupMetadata): GroupMetadata = {
+    groupsCache.put(groupId, group)
```
It's all coming back to me now! This is why CoordinatorMetadata had a ReentrantReadWriteLock with a simple mutable.HashMap:

Let's say two consumers of the same group(brand new, no GroupMetadata exists) do a JoinGroup at the same time. GroupCoordinator will notice that the group doesn't exist for thread1 and thread2 and will proceed to make a new GroupMetadata for each thread and one will overwrite the other's entry in the groupsCache!

hachikuji replied:
Couldn't we use putIfNotExists?

onurkaraman replied:
Yeah I guess.

guozhangwang replied:
Ack on this.

I am still trying to figure out if we still need the read-write lock on the group: one edge case that once the group metadata is fetched, its member's list can be modified by concurrent threads. Should this be requiring the global lock then? @onurkaraman @hachikuji

guozhangwang replied:
I think the group lock inside the `GroupCoordinator` is good enough for guarding 1) `removeGroup`, `addMember`, `removeMember` and update member metadata / assignments. So the current concurrency control should be sufficient. But let me know if there are still edge cases.

onurkaraman replied:
Right. Basically I think there are two things we need to watch out for:
1. making sure the GroupMetadata is consistent. This happens with the GroupCoordinator using the GroupMetadata's implicit lock on any sort of state access or modifications to the group.
2. making sure the mapping holding the GroupMetadata is consistent. This happens through careful usage of the Pool (like using Pool.putIfNotExists in addGroup).

guozhangwang committed:
change storeGroup to return error code upon latch awaiting

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+
+  private val CURRENT_OFFSET_SCHEMA_VERSION = 1.toShort
+
+  private val CURRENT_GROUP_SCHEMA_VERSION = 2.toShort
```
Do these refer to the value or key schemas?

guozhangwang replied:
Both. Currently the key / value schema versions are always the same. But moving forward they can be different.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,843 @@
+  private val GROUP_METADATA_MEMBERS_V0 = GROUP_METADATA_VALUE_SCHEMA_V0.get("members")
+
+  // map of versions to schemas
+  private val OFFSET_SCHEMAS = Map(0 -> KeyAndValueSchemas(OFFSET_COMMIT_KEY_SCHEMA_V0, OFFSET_COMMIT_VALUE_SCHEMA_V0),
```
Maybe rename OFFSET_SCHEMAS to MESSAGE_TYPES?

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -394,21 +405,21 @@ class GroupCoordinator(val brokerId: Int,
   def handleGroupImmigration(offsetTopicPartitionId: Int) = {
-    // TODO we may need to add more logic in KAFKA-2017
-    offsetManager.loadOffsetsFromLog(offsetTopicPartitionId)
+    groupManager.loadGroupsForPartition(offsetTopicPartitionId)
```
We will need to be a bit careful with when handleGroupImmigration() is called. Currently, we first change the leader (in makeLeader()) and then add the affected partition to loadingPartitions in loadGroupsForPartition(). In between, a HeartBeat request will see the coordinator, but see an empty group. Then an UNKNOWN_MEMBER_ID error will be returned incorrectly.

guozhangwang replied:
Ack. We will let the GroupMetadataManager itself maintain the owned partitions, which are going to be synced with the same loading partition object.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,846 @@
+      groupMetadataMessageSet,
+      putCacheCallback)
+
+    responseLatch.await()
```
This will tie up an I/O thread, which is not ideal.

guozhangwang replied:
Ack. Fixing it now, the originally motivation is that with the callback handling with parking callbacks the unit test `GroupCoordinatorResponseTest` becomes extremely tricky.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,846 @@
+  // map of versions to schemas
+  private val OFFSET_SCHEMAS = Map(0 -> KeyAndValueSchemas(OFFSET_COMMIT_KEY_SCHEMA_V0, OFFSET_COMMIT_VALUE_SCHEMA_V0),
+    1 -> KeyAndValueSchemas(OFFSET_COMMIT_KEY_SCHEMA_V0, OFFSET_COMMIT_VALUE_SCHEMA_V1),
+    2 -> KeyAndValueSchemas(GROUP_METADATA_KEY_SCHEMA_V0, GROUP_METADATA_VALUE_SCHEMA_V0))
```
In the map, it seems that the key of the OFFSET_SCHEMAS should be the version of the key schema.

guozhangwang replied:
As mentioned before, currently the versions of the key and value are always the same; moving forward we can have different key and value versions.

guozhangwang replied:
@junrao @hachikuji Just to make it more clear, I am going to update the schema maps to reflect the versions on keys only.

guozhangwang committed:
wip on fixing GroupCoordinatorResponseTest

guozhangwang committed:
address comments in github

onurkaraman commented:
I left a question on the original jira but it looks like it got lost among the updates here:

Regarding the schema evolution: what was the reason for making v1 represent offset schema and v2 represent group schema? I think we did a similar pattern for OffsetFetchRequest and OffsetCommitRequest (where one version represented commits to zookeeper and another represented commits to kafka), but I think circumstances were different in that kafka-based offset storage was meant to replace zookeeper-based offset storage.

guozhangwang commented:
@onurkaraman The reason for using key version as the message type indicator is that log compactor will compact different versions of the same data type (they will have different value version but the same key version), but not compact different data types together.

guozhangwang committed:
wip to use easyMock.capture

guozhangwang commented:
@junrao @hachikuji @onurkaraman Ping for another round of reviews.

Besides addressing comments the major changes are with `GroupCoordinatorResponseTest`: mokcing on the replica manager lever and adding EasyMock.capture for store offsets and groups.

guozhangwang committed:
fix one bug

guozhangwang commented:
Also created https://issues.apache.org/jira/browse/KAFKA-2720 to keep track of the periodic group purging.

guozhangwang committed:
include client id in member id, remove group in cache upon partition migration

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+  class OffsetsMessageFormatter extends MessageFormatter {
+    def writeTo(key: Array[Byte], value: Array[Byte], output: PrintStream) {
+      val formattedKey = if (key == null) "NULL" else GroupMetadataManager.readMessageKey(ByteBuffer.wrap(key)).toString
+      val formattedValue = if (value == null) "NULL" else GroupMetadataManager.readOffsetMessageValue(ByteBuffer.wrap(value)).toString
```
It seems that we need to support reading GroupMetadataValue too.

guozhangwang replied:
I was thinking it will be added in the patch of adding the describe-group request; I will coordinate with @hachikuji about this.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+
+  newGauge("NumGroups",
+    new Gauge[Int] {
+      def value = offsetsCache.keys.map(_.group).toSet.size
```
Could this be using groupsCache?

guozhangwang replied:
Oops, ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+
+  def partitionFor(group: String): Int = Utils.abs(group.hashCode) % groupMetadataTopicPartitionCount
+
+  def partitionLeaderIsLocal(partition: Int): Boolean = loadingPartitions synchronized ownedPartitions.contains(partition)
```
Perhaps we can replace partitionLeaderIsLocal() with a more intuitive api like isGroupCoordinatorLocal(group: String). Inside this method, we can use partitionFor to translate group to partition first. Then the caller only needs to deal with groups, not partitions.

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+
+  def partitionLeaderIsLocal(partition: Int): Boolean = loadingPartitions synchronized ownedPartitions.contains(partition)
+
+  def partitionLoadingInProgress(partition: Int): Boolean = loadingPartitions synchronized loadingPartitions.contains(partition)
```
Same as the above. Would it be better to replace this with a isGroupLoading(group: String)?

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+    val groupMetadataMessageSet = Map(groupMetadataPartition ->
+      new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, message))
+
+    // set the callback function to insert offsets into cache after log append completed
```
The comment seems incorrect since this is about consumer group metadata.

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+          .format(responseStatus, groupMetadataPartition))
+
+      // construct the commit response status and insert
+      // the offset and metadata to cache if the append status has no error
```
Inaccurate comment.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+      // the offset and metadata to cache if the append status has no error
+      val status = responseStatus(groupMetadataPartition)
+
+      var responseCode = Errors.NONE.code
```
responseCode doesn't seem to be used.

guozhangwang replied:
My bad, fixed.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+      }
+
+      for (member <- group.allMembers) {
+        member.assignment = groupAssignment.getOrElse(member.memberId, Array.empty[Byte])
```
Although storeGroup() is called with synchronization on group, member.assignment is updated in the callback and there doesn't seem to be a memory barrier. The same issue seems to exist in awaitingSyncCallback as well.

guozhangwang replied:
Ack. Adding volatile to these member fields.

hachikuji replied:
We've already transitioned to Stable, but the assignments will not be set until the callback is invoked. Is it possible for a follower SyncGroup to arrive before the callback is invoked?

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+      propagateAssignment(group, Errors.NONE.code)
+    }
+
+    // call replica manager to append the offset messages
```
Inaccurate comment.

guozhangwang replied:
Ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,919 @@
+    val topicPartition = TopicAndPartition(GroupCoordinator.GroupMetadataTopicName, offsetsPartition)
+
+    loadingPartitions synchronized {
+      ownedPartitions.add(offsetsPartition)
```
It seems that we should only add the partition to ownedPartitions after the loading completes.

guozhangwang replied:
I think this does not make a major difference here: with this approach when a partition is loading, coordinator responses will get GROUP_LOADING error code causing the client to retry; with the suggested approach the coordinator responses will get NOT_COORDINATOR_FOR_GROUP error code causing the client to re-discover and retry. I personally prefer the first approach as it does not introduce necessary re-discovering.

junrao commented:
1. When the last member leaves the group, we need to delete the group by writing the tombstone to the offset topic. So, it seems that in GroupCoordinator.onMemberFailure(), we should call GroupMetadataManager.removeGroup()?
2. Do we have unit tests that cover the loading of both the offsets and the groups metadata from the offset topic?

guozhangwang commented:
Rep. Jun 1): when `removeGroup()` is called on the last member of the group, it will first enter the "PrepareRebalance" state, and then directly trigger `onCompleteJoin` since the purgatory's delayed request can be completed immediately with no members. Hence we are effectively always call `removeGroup` as in `onCompleteJoin`. This code is just for making the state machine diagram more clear.

2). We have the unit tests for various coordinator responses that have been updated in this patch with `ReplicaManager.appendMessages` mock to test basic functions of `GroupMetadataManager`. For further integration tests that requires a full-fledged replica manager and coordinator / group manager we will include it in KAFKA-1326.

onurkaraman commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -458,10 +457,12 @@ class GroupCoordinator(val brokerId: Int,
                                     callback: JoinCallback) = {
-    val memberId = group.generateNextMemberId
+    // use the client-id with a random id suffix as the member-id
+    val memberId = clientId + "-" + group.generateMemberIdSuffix
```
What does this change have to do with persisting coordinator state for coordinator failover?

guozhangwang replied:
I think this is orthogonal to coordinator failover case: the client id is just used for easy debugging / tooling in addition to a global unique but random member id, it will just add a few bytes to the stored message but should be fine with Kafka.

onurkaraman replied:
Yeah I'm saying I kind of think this should be in a separate ticket.

guozhangwang replied:
Yeah there is a separate ticket KAFKA-2018, I was just trying to tackle it in this patch as well for shorter review cycle since it is small change.

guozhangwang committed:
address Jun's comments

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,939 @@
+  private val ownedPartitions: mutable.Set[Int] = mutable.Set()
+
+  /* lock for expiring stale offsets, it should be always called BEFORE the group lock if needed */
+  private val offsetExpireLock = new ReentrantReadWriteLock()
```
Does this need to be a readWriteLock? Both callers are using the writeLock. Perhaps this can just be a ReentrantLock?

guozhangwang replied:
ack.

junrao commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,939 @@
+  private val offsetExpireLock = new ReentrantReadWriteLock()
+
+  /* lock for removing offsets of a range partition, it should be always called BEFORE the group lock if needed */
+  private val offsetRemoveLock = new ReentrantReadWriteLock()
```
Do we need this lock? getOffsets() seems to already support the case when the offset is not present.

guozhangwang replied:
This lock is used to protect race condition when group migration happens at the same time during get offsets: if the group is removed after the `isGroupLocal` check, the coordinator with return empty offsets with NONE error code, instead of the right behavior of returning NOT_COORDINATOR error code, this will result in client side handling it as "we do not have any offsets stored in Kafka".

Actually I think this does expose a bug in the code: we need to include the removal of offsets also in the loadingPartition object lock, and hence we can remove this lock. @junrao Thoughts?

junrao commented:
Thanks for the patch. LGTM. Just a couple of minor comments.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -58,8 +58,8 @@
-    OFFSET_LOAD_IN_PROGRESS(14,
-            new OffsetLoadInProgressException("The coordinator is loading offsets and can't process requests.")),
+    GROUP_LOAD_IN_PROGRESS(14,
+            new OffsetLoadInProgressException("The coordinator is loading and hence can't process requests for this group.")),
```
Should we rename this exception?

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -29,7 +29,7 @@ import org.apache.kafka.common.requests.JoinGroupRequest
 
-case class GroupManagerConfig(groupMinSessionTimeoutMs: Int,
+case class GroupConfig(groupMinSessionTimeoutMs: Int,
                               groupMaxSessionTimeoutMs: Int)
```
nit: fix indentation

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,939 @@
+}
+
+
+object GroupMetadataManager {
```
Might be worth some high-level comments on the organization and versioning of the key/value schemas below, in particular in regard to the problems that must be addressed if we need to change one of the key versions.

guozhangwang replied:
ack.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,939 @@
+
+  private def addGroup(groupId: String, group: GroupMetadata): GroupMetadata = {
+    groupsCache.putIfNotExists(groupId, group)
+    group
```
Shouldn't we return the result of putIfNotExists?

guozhangwang replied:
Nice catch, though putIfNotExists will return the previous value of the key, hence it can be null. Instead we can return `groupsCache.get()` right after putIfNotExists.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala:
```diff
@@ -0,0 +1,939 @@
+    group.transitionTo(Dead)
+
+    if (groupsCache.remove(group.groupId) != group)
+      throw new IllegalArgumentException("Cannot remove group " + group.groupId + " since it has been replaced.")
```
This case would happen if two threads hit removeGroup simultaneously, right? If there's nothing preventing that case, would we be better returning rather than raising an exception?

hachikuji replied:
Nevermind. This is protected with the group lock.

hachikuji commented on core/src/main/scala/kafka/coordinator/GroupCoordinator.scala:
```diff
@@ -537,14 +533,14 @@ class GroupCoordinator(val brokerId: Int,
+        // TODO KAFKA-2720: only remove group in the background thread
         if (group.isEmpty) {
-          group.transitionTo(Dead)
+          groupManager.removeGroup(group)
```
I think the synchronization around adding/removing groups from the cache might not be sufficient. Suppose that this removeGroup() is hit while another thread is blocking on the group lock in doJoinGroup(). The group will be removed from the cache, but the other thread will still be able to acquire the lock when it is finished and perform operations on the group, right?

onurkaraman replied:
@hachikuji That's exactly why I added the Dead state.

The Dead state checks prevent you from acting on a GroupMetadata object that is in the process of being removed by another thread.

guozhangwang replied:
We have the "Dead" state in which removeGroup should always transit to within the lock, and we always check the state again after we grab the group lock. So I think it should be fine.

hachikuji replied:
Ah, yeah, that makes sense. I think I never got the reason for the dead check before.

guozhangwang committed:
address github comments round two
